docs/01-tooling/03-fast-and-isolated-JS-unit-tests.md

```diff
@@ -2,10 +2,159 @@
 
 ### Intro
 
+For testing frontend code, MediaWiki provides a browser based QUnit setup. For
+running the tests, you have to spin up MediaWiki, and load
+Special:JavaScriptTest in your browser, which will run all the QUnit tests for
+all the extensions and MediaWiki itself.
+
+From then on, you can filter by module or string, hide passed tests, and a few
+other things like select to load the assets in production mode or development
+mode (`debug=true`).
+
+Like any testing strategy, this setup comes with tradeoffs. Specifically, there
+are a couple that we have had problems with:
+
+#### Tests usually take a long time to run
+
+With this setup, they have to go through to the MediaWiki server,
+ResourceLoader, and then run in the browser. This is specially noticeable in
+development mode, which we often enable to get readable stack traces and error
+messages on test failures, but makes the test run take a lot longer.
+
+As a result of this, writing tests is costlier than it should, which discourages
+developers to write and run tests, and over time ends up affecting the quality
+of our code and QUnit test suites. It also puts significant barriers to TDD
+style workflows, which rely on constantly running tests and require a fast
+feedback cycle with the developer.
```

Comment:
There's a great history on this and one source is Refactoring: Improving the Design of Existing Code
by Martin Fowler. A couple of the principles of effective refactoring they discuss are: tests must be solid, tests must be fast, and tests should fail loudly and succeed quietly. In particular, this is a book about Java from 2002 that heavily emphasizes speed.

Reply:
Cool. This is one of those things that I'm having to explain explicitly because I'm not sure everyone understands or values it in the dev community. Let me know if you want me to add something.